docs/design.md

```diff
@@ -0,0 +1,273 @@
+
+## How It Works
+
+The high level architecture for Ansible provider is illustrated as below.
+
+![](images/ansible-provider-design.png)
+
+Ansible provider uses Kubernetes resource `AnsibleRun` to represent a certain Ansible run, e.g.: to run a set of Ansible roles or playbooks. The `AnsibleRun` resource also refers to a `ProviderConfig` resource including all information at provider level that are needed to support Ansible runs, e.g.: credentials needed to access a remote place that hosts Ansible contents, requirements needed to install as prerequisite before run Ansible contents. Regarding the Ansible contents, it can be collections, roles, or playbooks that are hosted in Ansible Galaxy, public or private Automation Hub, or GitHub repository.
+
+As the managed resource of Ansible provider, once the `AnsibleRun` resource is defined, the provider will download the Ansible contents into the provider working directory first via the method `Connect()` in AnsibleRun controller, then trigger the Ansible run according to the provider lifecycle via method `Observe()/Create()/Update()/Delete()`. 
+
+Inside the provider, it actually relies on the command line tool [`ansible-galaxy`](https://docs.ansible.com/ansible/latest/cli/ansible-galaxy.html) to download Ansible contents, then use [`ansible-runner`](https://ansible-runner.readthedocs.io/) to execute them. The module [`ansible-runner-http`](https://github.com/ansible/ansible-runner-http) is a plugin for ansible runner that allows to emit Ansible status and events to HTTP services in the form of POST events, so as to notify AnsibleRun controller about the run results and attatch that to `AnsibleRun` as status.
+
+### Working Directory
+
+The provider working directory is used to host the Ansible contents downloaded from remote place. It is currently inside the provider container so that will not be persisted permanently by the provider. As a result, when the provider pod restarts, the contents will be lost, but the provider will download them from remote place again.
+
+## Supported Sources
+
+There are two types of sources from which the Ansible contents can be retrieved, installed and run by Crossplane Provider for Ansible (or Ansible provider for short).
+
+### Inline
+
+This is maily for quick test. You can inline the Ansible content to be run in an `AnsibleRun` resource. It will be wrapped as a `playbook.yml` file and stored in the working directory for the provider to run.
+
+Here is an example to call a builtin Ansible module:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: inline-example
+spec:
+  forProvider:
+    # For simple cases, you can use an inline source to specify the content of
+    # playbook.yml as opaque and inline yaml.
+    playbookInline: |
```

Comment:
I think we should have source field to support multiple source ( inline and remote)
```
    source: Inline
    module: |
      ---
      - hosts: localhost
        tasks:
          - name: ansibleplaybook-simple
            debug:
              msg: Your are running 'ansibleplaybook-simple' example
```

Reply:
yeah, actually I was doubting whether we could simplify it by presenting the notion of source implicitly. A couple of reasons:
- Given that we define `requirements` in `ProviderConfig`, any remote need can be expressed in `ProviderConfig` instead of in `AnsibleRun`
- For inline playbook, it has already been covered by `spec. forProvider.playbookInline`
- For inline + remote case, again, the remote need has been expressed in `ProviderConfig`, so we only need to define the inline playbook by reusing the field `spec. forProvider.playbookInline`.

With that, it looks we never need the additional `source` field.

Does it make sense.

Reply:
Re: the last point.

> not all remote ansible projects are collections so not all remote ansible project should be fetched as a requirement, e.g a simple ansible github project should be cloned with go-getter.

When I was reading the Ansible docs, I was doubting this as I see requirements.yml can support both collections and roles, e.g.:

```yaml
---
roles:
  # Install a role from Ansible Galaxy.
  - name: sample_namespace.sample_role
    version: 0.1.0
collections:
  # Install a collection from Ansible Galaxy.
  - name: sample_namespace.sample_collection
    version: 0.1.0
    source: https://galaxy.ansible.com
```

It can even support roles retrieved from git repo like this:

```yaml
---
roles:
  - src: git@github.com:sample_namespace/sample_collection.git
    scm: git
    version: master
```

With that, it looks we can fully rely on requirements.yml and ansible-galaxy, which can simplify both spec and implementation. Did I miss anything here?

Reply:
@morningspace  I agree technically it is possible to rely on `requirements.yml` to fetch remote ansible project.. But it is important to separate the dependencies from the core project. Functionality speak when  applying an ansibleRun resource we aim to run an ansible project which can have dependencies. So here the question is to keep the logical separation between a project and its dependencies.

Reply:
Ah, I see your point. We differentiate the "project" and its dependencies. For the project itself, it should not be put in `requirement.yml`, agree. I will make the change. For the implementation details, I think it may still be viable to use `ansible-galaxy` to retrieve both the "project" and its dependencies, because `ansible-galaxy` supports multiple sources, such as Ansible Galaxy, Automation Hub, etc. not just Git repo.

Reply:
Yes  the use of  ansible-galaxy to fetch project is more efficient then a simple getter 👍

Reply:
I see your point :-) .

from a functional point of view redundancy is legitimate. Indeed, the goal that the user can express the difference between the project and its dependencies through the spec. But technically, it may be a bit heavy. Now we will try to keep it simple for the first release. the actual code make the separation between these two concerns. 

the the requirement fields I think we should keep it for consistency with ansible terminologies